runtime/bindings/python/hal.cc

```diff
@@ -411,54 +411,59 @@ HalDevice HalDriver::CreateDeviceByURI(std::string& device_uri,
 namespace {
 
 py::object MapElementTypeToDType(iree_hal_element_type_t element_type) {
-  // See: https://docs.python.org/3/c-api/arg.html#numbers
-  // TODO: Handle dtypes that do not map to a code (i.e. fp16).
-  const char* dtype_code;
+  // See:
+  //   * https://numpy.org/doc/stable/reference/arrays.dtypes.html
+  //   * https://docs.python.org/3/c-api/arg.html#numbers
+  //
+  // Single letter codes can be ambiguous across platforms, so prefer explicit
+  // bit depth values, ("Type strings: Any string in numpy.sctypeDict.keys()").
+  // See https://github.com/pybind/pybind11/issues/1908
+  const char* dtype_string;
```

Comment:
Hey @Eliasj42 , this is changing the same code as your https://github.com/openxla/iree/pull/12872 (nice timing on that and @allieculp 's ping on the Windows issue!). I think adding a complex test case to runtime/bindings/python/tests/vm_types_test.py might be enough for your PR, but we should also add more e2e tests using complex to ensure it works across all the layers.